### PR TITLE
Fix some asmap issues

### PR DESCRIPTION
**Pull request by sipa**
Here are a few things to improve in the asmap implementation. The first two commits are just code improvements. The last one is a bugfix (the exsting code wouldn't correctly apply ASN lookups to mapped/embedded IPv4 addresses).

### CODE REVIEW

File: src/netaddress.h [hunk 1]
```diff
@@ -80,6 +80,11 @@ class CNetAddr
         bool GetInAddr(struct in_addr* pipv4Addr) const;
         uint32_t GetNetClass() const;
 
+        //! For IPv4, mapped IPv4, SIIT translated IPv4, Teredo, 6to4 tunneled addresses, return the relevant IPv4 address as a uint32.
+        bool HasLinkedIPv4() const;
+        //! Whether this address has a linked IPv4 address (see GetLinkedIPv4()).
```
Comment (practicalswift) [inline src/netaddress.h]: This comment should be on L83 and vice versa? :)

**Comment by fanquake** [comment]
cc @naumenkogs

File: src/netaddress.h (hunk 1 above)
Comment (sipa) [reply src/netaddress.h]: Fixed.

**Comment by sanjaykdragon** [comment]
utack - looks nice, preventing unnecessary copies

**Comment by practicalswift** [comment]
Concept ACK

While improving asmap -- please consider increasing the robustness of `GetMappedAS(…)` by plugging this issue too: ["GetMappedAS(...) lookup on an IPv6 address against a maliciously constructed AS-map triggers heap buffer-overflow"](https://github.com/bitcoin/bitcoin/issues/18033) :)

**Comment by sipa** [comment]
Added a commit to resolve #18033.

**Comment by jonatack** [comment]
Concept ACK, will review shortly.

**Comment by naumenkogs** [comment]
utACK 868647e5d72f2718e41b723dc7e15594ffb1d7b6
Thank you!

File: src/util/asmap.cpp [hunk 2]
```diff
@@ -35,52 +37,53 @@ uint32_t DecodeBits(std::vector<bool>::const_iterator& bitpos, uint8_t minval, c
...
-        opcode = DecodeType(pos);
+    while (pos != endpos) {
+        opcode = DecodeType(pos, endpos);
         if (opcode == 0) {
-            return DecodeASN(pos);
+            return DecodeASN(pos, endpos);
         } else if (opcode == 1) {
-            jump = DecodeJump(pos);
+            jump = DecodeJump(pos, endpos);
             if (ip[ip.size() - bits]) {
```
Comment (practicalswift) [inline src/util/asmap.cpp]: A heap buffer overflow will occur here in case of `bits == 0`?

File: src/util/asmap.cpp [hunk 3]
```diff
@@ -35,52 +37,53 @@ uint32_t DecodeBits(std::vector<bool>::const_iterator& bitpos, uint8_t minval, c
...
 
 }
 
 uint32_t Interpret(const std::vector<bool> &asmap, const std::vector<bool> &ip)
 {
     std::vector<bool>::const_iterator pos = asmap.begin();
+    const std::vector<bool>::const_iterator endpos = asmap.end();
     uint8_t bits = ip.size();
-    uint8_t default_asn = 0;
+    uint32_t default_asn = 0;
```
Comment (practicalswift) [inline src/util/asmap.cpp]:
```suggestion
    uint32_t default_asn = 0;
    if (bits == 0) {
        return default_asn;
    }
```

File: src/util/asmap.cpp (hunk 2 above)
Comment (practicalswift) [inline src/util/asmap.cpp]:
```suggestion
             if (bits == 0) {
                 return default_asn;
             }
             if (ip[ip.size() - bits]) {
```

File: src/util/asmap.cpp [hunk 4]
```diff
@@ -35,52 +37,53 @@ uint32_t DecodeBits(std::vector<bool>::const_iterator& bitpos, uint8_t minval, c
...
+                if (jump >= endpos - pos) break;
                 pos += jump;
             }
             bits--;
         } else if (opcode == 2) {
-            match = DecodeMatch(pos);
+            match = DecodeMatch(pos, endpos);
             matchlen = CountBits(match) - 1;
             for (uint32_t bit = 0; bit < matchlen; bit++) {
                 if ((ip[ip.size() - bits]) != ((match >> (matchlen - 1 - bit)) & 1)) {
```
Comment (practicalswift) [inline src/util/asmap.cpp]:
```suggestion
                 if (bits == 0 || (ip[ip.size() - bits]) != ((match >> (matchlen - 1 - bit)) & 1)) {
```

File: src/util/asmap.cpp (hunk 2 above)
Comment (sipa) [reply src/util/asmap.cpp]: No, reaching the end of the ip array just means a malformed asmap. It's probably better to return 0 (also the other location).

**Comment by sipa** [comment]
@practicalswift Nice catch, addressed I think.

File: src/util/asmap.cpp (hunk 2 above)
Comment (sipa) [reply src/util/asmap.cpp]: Yup, fixed.

**Comment by practicalswift** [comment]
@sipa I'm no longer able to trigger the two heap buffer overflows, and the code withstood an eight hour `asmap` fuzzer (#18033) run (48 million executions): I think the code is robust against malicious map data now :)

ACK c86bc144081f960347232546f7d22deb65d27deb -- patch looks correct

File: src/netaddress.cpp [hunk 5]
```diff
@@ -463,51 +497,32 @@ std::vector<unsigned char> CNetAddr::GetGroup(const std::vector<bool> &asmap) co
...
-    }
-    else if (IsTor())
-    {
+    } else if (IsTor()) {
         nStartByte = 6;
         nBits = 4;
-    }
-    // for he.net, use /36 groups
-    else if (GetByte(15) == 0x20 && GetByte(14) == 0x01 && GetByte(13) == 0x04 && GetByte(12) == 0x70)
+    } else if (GetByte(15) == 0x20 && GetByte(14) == 0x01 && GetByte(13) == 0x04 && GetByte(12) == 0x70) {
```
Comment (jonatack) [inline src/netaddress.cpp]: This line could be extracted to a `bool CNetAddr::IsHeNet() const` function

**Review by jonatack — COMMENTED** [review]
ACK c86bc144081f960347232546f7d22deb65d27deb code looks correct, built/ran tests, bitcoind with -asmap pointed to asmap/demo.map

Question: is it expected to see as many identical values?
```
$ bitcoin-cli getpeerinfo | grep 'mapped_as' | sort
    "mapped_as": 1267,
    "mapped_as": 12876,
    "mapped_as": 14061,
    "mapped_as": 15435,
    "mapped_as": 202053,
    "mapped_as": 20473,
    "mapped_as": 24940,
    "mapped_as": 24940,
    "mapped_as": 30083,
    "mapped_as": 31334,
    "mapped_as": 34878,
    "mapped_as": 34878,
    "mapped_as": 39586,
    "mapped_as": 4713,
    "mapped_as": 47869,
    "mapped_as": 54098,
    "mapped_as": 54098,
    "mapped_as": 54098,
    "mapped_as": 54098,
    "mapped_as": 54098,
```

Edit: am fuzzing this PR with #18029; currently at 34.6M executions.
```
#34611198 REDUCE cov: 1198 ft: 3529 corp: 197/13577b exec/s: 1125 rss: 450Mb ...
```

**Comment by jonatack** [comment]
On my side, this PR held up on 72M execs (about 18 hours) of the asmap fuzzer.
```
#72474660	REDUCE cov: 1198 ft: 3529 corp: 197/12792b exec/s: 1063 rss: 452Mb L: 41/1527 MS: 4 InsertByte-EraseBytes-InsertByte-PersAutoDict- DE: "\x00 \x00\x00\x00\x00\x00\x00"-
```

**Comment by naumenkogs** [comment]
utACK c86bc14

**Comment by laanwj** [comment]
ACK c86bc144081f960347232546f7d22deb65d27deb

**Pull request merged**